docs/layer-basic.md

```diff
@@ -116,6 +125,18 @@ This layer supports the following options, which can be set in `layer.yaml`:
     the `--system-site-packages` options to make system Python libraries
     visible within the venv.
 
+  * **autoremove_config_flags** If set to true, the legacy behavior of
+    automatically removing the various `config.*` flags is enabled.  Otherwise,
+    the flags will not be removed and they should be removed by the top-level
+    charm layer as processed, while base layers should use [triggers].
```

Comment:
> I don't think that shifting the burden on clearing framework maintained flags onto the charm improves things. 

I actually think it improves things a lot. It's important to reduce the amount of "magic" as much as possible. Moreover, the "split-brain" issue will exist as long as we require charmers to think about _both_ charms.reactive and the hook lifecycle while writing handlers.

Granted, people who are familiar with hooks will need to learn this mapping, but that is a one-time effort and not a big one if we explain the behavior clearly. Especially since this isn't that far from the hook paradigm; most hooks just signal that something changed. When the charms.reactive framework gets such a signal, it sets a flag.

Look at the conversations -> endpoints shift. Even though conversations are more powerful, many people got stuck on them because their behavior is different depending on what hook is running.  In order to use conversations successfully, you have to be an expert in how relationships and their hooks work. Endpoints provide a lot less functionality, but what they do provide is reliable and consistent. Regardless of the hook you're in, everything will work the same and you only need to understand the flags in order to be able to use them.

Conversations, just like with the `config.changed` flags, also have the issue that it's very easy to write wrong code that _seemingly_ works but has unexpected behavior in edge cases. If a handler reacts to multiple preconditions, one of them being a relation state, then your charm might- or might not crash depending on whether the other preconditions are true _before-_ or _after_ the relationship hook during which the relation state is set. It's the same with config changed, although not so dramatically: a handler reacting to `config.changed` might or might not run depending on whether its other preconditions match during the `config-changed` hook or after. Moreover, most of the times when a charm _does_ rely on this behavior, it's still incorrect. For example, if you want to restart a server only when the config changes _after_ the server is started:

```python
@when('config.set.x')
def start():
    # generate config
    # start server
    set_flag('server.running')

@when('config.changed.x', 'server.running')
def restart():
    # update config
    # restart server
```

This will restart the server immediately after the first start. This is a typical beginners mistake that is probably present in many charms used in production. An easy fix is to add `clear_flag('config.changed.x')` to the `start` handler, but then we're already manually clearing these flags.. So just quit the magic, even if it means writing one more line of code..